### PR TITLE
Limit file upload size

### PR DESCRIPTION
**Pull request by MarioRodrigues10**

### CODE REVIEW

File: lib/bokken/uploaders/avatar.ex [hunk 1]
```diff
@@ -7,13 +7,28 @@ defmodule Bokken.Uploaders.Avatar do
 
   @versions [:original, :thumb]
   @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @min_file_size 10_000
```
Comment (ruilopesm) [inline lib/bokken/uploaders/avatar.ex]: I don't think there should be a minimum file size.

File: lib/bokken/uploaders/avatar.ex [hunk 2]
```diff
@@ -7,13 +7,28 @@ defmodule Bokken.Uploaders.Avatar do
...
+      size < @min_file_size ->
+        {:error, "File is too small. Minimum size is #{@min_file_size} bytes."}
+
+      size > @max_file_size ->
+        {:error, "File is too large. Maximum size is #{@max_file_size} bytes."}
+
+      true ->
+        {:ok, file}
+    end
   end
```
Comment (ruilopesm) [inline lib/bokken/uploaders/avatar.ex]: As this is now, if the file doesn't meet the `@extension_whitelist`, but meets the file size requirements it is taken as a valid file. Remember that Elixir only returns the last evaluated expression.

File: lib/bokken/uploaders/avatar.ex [hunk 3]
```diff
@@ -37,4 +52,9 @@ defmodule Bokken.Uploaders.Avatar do
   def default_url(_version, scope) do
     "https://robohash.org/#{scope.first_name}-#{scope.last_name}"
   end
+
+  def file_size(%Waffle.File{} = file) do
```
Comment (ruilopesm) [inline lib/bokken/uploaders/avatar.ex]:
```suggestion
  defp file_size(%Waffle.File{} = file) do
```

File: lib/bokken/documents/file.ex [hunk 4]
```diff
@@ -40,4 +43,9 @@ defmodule Bokken.Documents.File do
     |> validate_required(@required_fields ++ @attachment_fields)
     |> assoc_constraint(:user)
   end
+
+  def file_size(file) do
+    File.stat!(file.path)
+    |> Map.get(:size)
+  end
```
Comment (ruilopesm) [inline lib/bokken/documents/file.ex:49]: Why is this function needed here?

File: lib/bokken/uploaders/avatar.ex [hunk 5]
```diff
@@ -7,13 +7,23 @@ defmodule Bokken.Uploaders.Avatar do
 
   @versions [:original, :thumb]
   @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @max_file_size 50_000_000
```
Comment (ruilopesm) [inline lib/bokken/uploaders/avatar.ex]: Don't you think 50MB is too much? 😅
An avatar should have a cap of 2MB, for example.
@ruioliveira02 What's your take regarding this?

File: lib/bokken/uploaders/avatar.ex [hunk 6]
```diff
@@ -7,13 +7,23 @@ defmodule Bokken.Uploaders.Avatar do
...
-    |> then(&Enum.member?(@extension_whitelist, &1))
+    |> then(&Enum.member?(@extension_whitelist, &1)) and check_file_size(size)
+  end
+
+  defp check_file_size(size) do
+    if size > @max_file_size do
+      false
+    else
+      true
+    end
```
Comment (ruilopesm) [inline lib/bokken/uploaders/avatar.ex]: You can remove this function and simply do `and size > @max_file_size` above. I think it's prettier and saves a function call.

File: lib/bokken/uploaders/document.ex [hunk 7]
```diff
@@ -4,8 +4,26 @@ defmodule Bokken.Uploaders.Document do
   """
   use Waffle.Definition
   use Waffle.Ecto.Definition
+  @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @max_file_size 50_000_000
```
Comment (ruilopesm) [inline lib/bokken/uploaders/document.ex]: Same here. I imagine someone uploading a Python script, for example. Those are not too heavy, as they are just text.

File: lib/bokken/uploaders/document.ex [hunk 8]
```diff
@@ -4,8 +4,26 @@ defmodule Bokken.Uploaders.Document do
...
+
+    file.file_name
+    |> Path.extname()
+    |> String.downcase()
+    |> then(&Enum.member?(@extension_whitelist, &1)) and check_file_size(size)
+  end
+
+  defp check_file_size(size) do
+    if size > @max_file_size do
+      {:error, "File size is too large"}
```
Comment (ruilopesm) [inline lib/bokken/uploaders/document.ex]: This is not doing nothing here. Elixir returns the last evaluated value.

File: lib/bokken_web/controllers/file_controller.ex [hunk 9]
```diff
@@ -55,12 +55,25 @@ defmodule BokkenWeb.FileController do
...
+    total_size = Enum.reduce(files, 0, fn file, acc -> acc + File.file_size(file) end)
 
     with {:ok, %File{} = file} <- Documents.create_file(Map.put(file_params, "user_id", user_id)) do
-      conn
-      |> put_status(:created)
-      |> put_resp_header("location", Routes.file_path(conn, :show, file))
-      |> render("show.json", file: file)
+      file_size = File.file_size(file)
+
+      if total_size + file_size > 100_000_000 do
```
Comment (ruilopesm) [inline lib/bokken_web/controllers/file_controller.ex]: 100MB for user is too much. I would suggest 6MB.

File: lib/bokken/uploaders/document.ex [hunk 10]
```diff
@@ -4,8 +4,18 @@ defmodule Bokken.Uploaders.Document do
   """
   use Waffle.Definition
   use Waffle.Ecto.Definition
+  alias Bokken.Documents.File
+  @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @max_file_size 5_000_000
```
Comment (ruilopesm) [inline lib/bokken/uploaders/document.ex]:
```suggestion
  @max_file_size 4_000_000
```

If the maximum is 6MB, it doesn't make much sense to put 2 for an avatar and 5 for a document (2 + 5 = 7 > 6).

File: lib/bokken/uploaders/emblem.ex [hunk 11]
```diff
@@ -4,16 +4,19 @@ defmodule Bokken.Uploaders.Emblem do
   """
   use Waffle.Definition
   use Waffle.Ecto.Definition
+  alias Bokken.Documents.File
 
   @versions [:original]
   @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @max_file_size 5_000_000
```
Comment (ruilopesm) [inline lib/bokken/uploaders/emblem.ex]:
```suggestion
  @max_file_size 1_000_000
```

1MB is enough for this. These are uploaded by us, so that, we can control it 🙏🏼

**Review by ruilopesm — COMMENTED** [review]
Can you implement tests for this? I think it's an important feature and should have it.

File: lib/bokken/documents/file.ex [hunk 12]
```diff
@@ -4,6 +4,8 @@ defmodule Bokken.Documents.File do
   """
   use Bokken.Schema
   use Waffle.Ecto.Schema
```
Comment (nelsonmestevao) [inline lib/bokken/documents/file.ex]: It is duplicated in `Bokken.Schema` :eyes:

**Comment by MarioRodrigues10** [comment]
Testing is failed due to ```EnrollmentControllerTest ```, ```EventsTest``` and ```AvailabilityControllerTest``` so it isn't related to this PR.

File: test/bokken/accounts_test.exs [hunk 13]
```diff
@@ -112,4 +113,120 @@ defmodule Bokken.AccountsTest do
...
+    test "delete_mentor/1 deletes the mentor" do
+      mentor = mentor_fixture()
+      assert {:ok, %Mentor{}} = Accounts.delete_mentor(mentor)
+      assert_raise Ecto.NoResultsError, fn -> Accounts.get_mentor!(mentor.id) end
+    end
+
+    test "change_mentor/1 returns a mentor changeset" do
+      mentor = mentor_fixture()
+      assert %Ecto.Changeset{} = Accounts.change_mentor(mentor)
+    end
```
Comment (ruilopesm) [inline test/bokken/accounts_test.exs]: I think this should be related to another PR.

Comment (MarioRodrigues10) [reply test/bokken/accounts_test.exs]: You're right

File: lib/bokken_web/controllers/file_controller.ex [hunk 14]
```diff
@@ -55,12 +55,25 @@ defmodule BokkenWeb.FileController do
...
 
     with {:ok, %File{} = file} <- Documents.create_file(Map.put(file_params, "user_id", user_id)) do
-      conn
-      |> put_status(:created)
-      |> put_resp_header("location", Routes.file_path(conn, :show, file))
-      |> render("show.json", file: file)
+      file_size = File.file_size(file)
+
+      if total_size + file_size > 6_000_000 do
+        Documents.delete_file(file)
```
Comment (ruilopesm) [inline lib/bokken_web/controllers/file_controller.ex]: This check should be inside `create_file` function. Imagine a case where the `delete_file` returns an error - we would stay with a created file.
I would suggest you to put some kinda of check on the changeset used by the `create_file` function. Since you also pass `user_id` to the function, it would be very easy to add a check on the changeset pipeline.

Ideally the _file size per user_ would be a column on the user's table, but we can take that to another PR.

File: lib/bokken_web/controllers/file_controller.ex [hunk 15]
```diff
@@ -55,12 +55,25 @@ defmodule BokkenWeb.FileController do
...
+
+        conn
+        |> put_status(:forbidden)
+        |> render("error.json", message: "File size is too large")
```
Comment (ruilopesm) [inline lib/bokken_web/controllers/file_controller.ex]:
```suggestion
        |> render("error.json", message: "You exceeded the maximum storage quota. Try to delete one or more files")
```

File: lib/bokken/documents.ex [hunk 16]
```diff
@@ -117,4 +121,11 @@ defmodule Bokken.Documents do
...
     File.changeset(file, attrs)
   end
+
+  def verify_total_size(file, user_id) do
+    user = Accounts.get_user!(user_id) |> Repo.preload(:files)
+
+    total_size = Enum.reduce(user.files, 0, fn file, acc -> acc + File.file_size(file) end)
+    total_size + File.file_size(file)
+  end
 end
```
Comment (ruioliveira02) [inline lib/bokken/documents.ex]: This should be a column in the user table. This way, you don't have to perform a full table scan on the files table whenever a user uploads a new file

**Review by ruioliveira02 — CHANGES_REQUESTED** [review]

File: lib/bokken/documents.ex (hunk 16 above)
Comment (MarioRodrigues10) [reply lib/bokken/documents.ex]: Considering that the size and quantity of files per user are not very large, does it make sense to store the size in the database? Performing that function for each user is extremely lightweight, while putting it in the database can generate other issues.

**Review by ruioliveira02 — COMMENTED** [review]
Solve the conflict, make sure the CI passes. Also, when you can, could you write a script as to help migrate this in production?

File: lib/bokken/uploaders/document.ex [hunk 17]
```diff
@@ -4,8 +4,18 @@ defmodule Bokken.Uploaders.Document do
   """
   use Waffle.Definition
   use Waffle.Ecto.Definition
+  alias Bokken.Documents.File
+  @extension_whitelist ~w(.jpg .jpeg .gif .png)
```
Comment (ruioliveira02) [inline lib/bokken/uploaders/document.ex]: We need to allow more extensions to be saved. In fact I think it would be easier to ban certain extensions (anything that is an executable file for example) then it would be to have a huge list of allowed extensions

File: lib/bokken/uploaders/document.ex [hunk 18]
```diff
@@ -4,8 +4,18 @@ defmodule Bokken.Uploaders.Document do
...
+  @max_file_size 4_000_000
```
Comment (ruioliveira02) [inline lib/bokken/uploaders/document.ex:10]: This should not be hardcoded

File: lib/bokken/uploaders/emblem.ex [hunk 19]
```diff
@@ -4,16 +4,19 @@ defmodule Bokken.Uploaders.Emblem do
   """
   use Waffle.Definition
   use Waffle.Ecto.Definition
+  alias Bokken.Documents.File
 
   @versions [:original]
   @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @max_file_size 1_000_000
```
Comment (ruioliveira02) [inline lib/bokken/uploaders/emblem.ex:11]: Same here

File: lib/bokken/uploaders/avatar.ex [hunk 20]
```diff
@@ -4,16 +4,19 @@ defmodule Bokken.Uploaders.Avatar do
   """
   use Waffle.Definition
   use Waffle.Ecto.Definition
+  alias Bokken.Documents.File
 
   @versions [:original, :thumb]
   @extension_whitelist ~w(.jpg .jpeg .gif .png)
+  @max_file_size 2_000_000
```
Comment (ruioliveira02) [inline lib/bokken/uploaders/avatar.ex:11]: Don't hardcode these values

File: lib/bokken/documents.ex [hunk 21]
```diff
@@ -86,10 +86,26 @@ defmodule Bokken.Documents do
...
+      |> File.changeset(attrs)
+      |> Repo.insert()
+
+      user = Accounts.get_user!(attrs.user_id)
+
+      user
+      |> User.changeset(%{total_file_size: verify_total_size(attrs.document, attrs.user_id)})
+      |> Repo.update()
+
+      {:ok, %File{}}
```
Comment (ruioliveira02) [inline lib/bokken/documents.ex]: 1. This does not return a useful result. It is a constant function after all
2. This must be enclosed within a transaction, so that either both operations work or both fail. See [Ecto.Multi](https://hexdocs.pm/ecto/Ecto.Multi.html)

File: lib/bokken/documents.ex [hunk 22]
```diff
@@ -138,4 +154,10 @@ defmodule Bokken.Documents do
   def change_file(%File{} = file, attrs \\ %{}) do
     File.changeset(file, attrs)
   end
+
+  def verify_total_size(file, user_id) do
+    user = Accounts.get_user!(user_id)
+
+    user.total_file_size + File.file_size(file)
+  end
```
Comment (ruioliveira02) [inline lib/bokken/documents.ex]: This name is misleading. It should be `get_new_total_size` or something like that.

**Review by ruioliveira02 — CHANGES_REQUESTED** [review]

File: lib/bokken_web/controllers/file_controller.ex [hunk 23]
```diff
@@ -52,7 +52,7 @@ defmodule BokkenWeb.FileController do
   def create(conn, %{"file" => file_params}) do
     user_id = conn.assigns.current_user.id
 
-    with {:ok, %File{} = file} <- Documents.create_file(Map.put(file_params, "user_id", user_id)) do
+    with {:ok, {:ok, file}} <- Documents.create_file(Map.put(file_params, "user_id", user_id)) do
```
Comment (ruioliveira02) [inline lib/bokken_web/controllers/file_controller.ex]: You can't assume the call will be successful

File: lib/bokken/documents.ex [hunk 24]
```diff
@@ -86,10 +86,57 @@ defmodule Bokken.Documents do
...
+      changeset = File.changeset(%File{}, map)
+
+      case Repo.insert(changeset) do
+        {:ok, file} ->
+          user_changeset =
+            User.changeset(Accounts.get_user!(map[:user_id]), %{
+              total_file_size: total_size
+            })
+
+          update_user_withchangeset(user_changeset, file)
```
Comment (ruilopesm) [inline lib/bokken/documents.ex]:
```suggestion
          update_user_with_changeset(user_changeset, file)
```

File: lib/bokken/documents.ex [hunk 25]
```diff
@@ -86,10 +86,57 @@ defmodule Bokken.Documents do
...
+
+        {:error, _} ->
+          {:error, "Failed to create file"}
+      end
+    end)
+  end
+
+  defp update_user_withchangeset(user_changeset, file) do
```
Comment (ruilopesm) [inline lib/bokken/documents.ex]:
```suggestion
  defp update_user_with_changeset(user_changeset, file) do
```

File: lib/bokken/documents.ex [hunk 26]
```diff
@@ -86,10 +86,57 @@ defmodule Bokken.Documents do
       {:error, %Ecto.Changeset{}}
 
   """
+  alias Bokken.Accounts.User
+
   def create_file(attrs \\ %{}) do
-    %File{}
-    |> File.changeset(attrs)
-    |> Repo.insert()
```
Comment (ruilopesm) [inline lib/bokken/documents.ex]: You also need this function case for things like seeds, for example.

File: lib/bokken/documents.ex [hunk 27]
```diff
@@ -75,23 +75,76 @@ defmodule Bokken.Documents do
...
+  def create_file_with_seeds(attrs \\ %{}) do
     %File{}
     |> File.changeset(attrs)
     |> Repo.insert()
   end
 
+  def create_file(attrs \\ %{}) do
+    case Application.fetch_env!(:bokken, Bokken.Uploaders.Document)[:max_file_size] do
+      nil ->
+        {:error, "MAX_TOTAL_SIZE environment variable not defined"}
```
Comment (ruilopesm) [inline lib/bokken/documents.ex]: You can remove this check.

File: lib/bokken/documents.ex [hunk 28]
```diff
@@ -75,23 +75,76 @@ defmodule Bokken.Documents do
...
 
-      iex> create_file(%{field: bad_value})
+      iex> create_file_with_seeds(%{field: bad_value})
       {:error, %Ecto.Changeset{}}
 
   """
-  def create_file(attrs \\ %{}) do
+  alias Bokken.Accounts.User
+
+  def create_file_with_seeds(attrs \\ %{}) do
```
Comment (ruilopesm) [inline lib/bokken/documents.ex]: The function should also be called `create_file`. It's just another case, not another function.

**Review by ruioliveira02 — APPROVED** [review]
